Jint/Runtime/Debugger/CallFrame.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using Esprima;
+using Esprima.Ast;
+using Jint.Native;
+using Jint.Runtime.CallStack;
+using Jint.Runtime.Environments;
+
+namespace Jint.Runtime.Debugger
+{
+    public class CallFrame
+    {
+        private readonly ExecutionContext _context;
+        private readonly CallStackElement? _element;
+        private readonly Lazy<DebugScopes> _scopeChain;
+
+        private LexicalEnvironment Environment => _context.LexicalEnvironment;
+
+        // TODO: CallFrameId
+        /// <summary>
+        /// Name of the function of this call frame. For global scope, this will be "(anonymous)".
+        /// </summary>
+        public string FunctionName => _element?.ToString() ?? "(anonymous)";
+
+        /// <summary>
+        /// Source location of function of this call frame.
+        /// </summary>
+        /// <remarks>For top level (global) call frames, as well as functions not defined in script, this will be null.</remarks>
+        public Location? FunctionLocation => (_element?.Function._functionDefinition?.Function as Node)?.Location;
+
+        /// <summary>
+        /// Currently executing source location in this call frame.
+        /// </summary>
+        public Location Location { get; }
+
+        /// <summary>
+        /// The scope chain of this call frame.
+        /// </summary>
+        public DebugScopes ScopeChain => _scopeChain.Value;
+
+        /// <summary>
+        /// The value of <c>this</c> in this call frame.
+        /// </summary>
+        public JsValue This => GetThis();
+
+        /// <summary>
+        /// The return value of this call frame. Will be null for call frames that aren't at the top of the stack,
+        /// as well as if execution is not at a return point.
+        /// </summary>
+        public JsValue ReturnValue { get; }
+
+        internal CallFrame(CallStackElement? element, ExecutionContext context, Location location, JsValue returnValue)
+        {
+            _element = element;
+            _context = context;
+            Location = location;
+            ReturnValue = returnValue;
+
+            _scopeChain = new Lazy<DebugScopes>(() => new DebugScopes(Environment));
+        }
+
+        private JsValue GetThis()
```

Comment:
looks a lot like `Engine.GetThisEnvironment`, could these share logic somehow?

Reply:
Would it be OK to move `GetThisEnvironment` to `LexicalEnvironment` (since `CallFrame` needs it for different execution contexts), or keep it in `Engine` but add an optional `LexicalEnvironment` parameter?

Reply:
Fine to move.  Only name is important as it's referred to in the spec algorithms.

Reply:
Moved to `ExecutionContext`. I've changed calls to call `(_engine).ExecutionContext.GetThisEnvironment` minimizing method calls - but let me know if you prefer a stub `Engine.GetThisEnvironment` calling `ExecutionContext.GetThisEnvironment` instead (guessing it can be inlined anyway).

That should be all in terms of addressing the remarks - although there's plenty of new stuff to remark on now. 😆